### PR TITLE
Tracing Re-Work and Context Support

### PR DESCRIPTION
**Pull request by iamemilio**
This is a very serious re-structuring of the way that tracing information is passed and stored, allowing us to move logic related to it into a centralized code location. This lets us define our situational policy for how tracing gets handled and passed in one place, and then invoke that library like an SDK to access tracing throughout the application. This should enable much more easy tracing for libraries that we integrate moving forward, and will substantially reduce the bug burden in the long run.

### Core changes
1. A lot of code was moved into common libraries where it is able to be tested and safely re-used
2. Boutique information around how tracing is passed has been moved to its own library/tool
3. The TraceFunction function now handles passing tracing to downstream calls, creating segments, and pulling the transaction out of a parameter

### CODE REVIEW

**Comment by iamemilio** [comment]
This is getting really, big so I am making sure that the unit tests to cover these changes are in absolute tip top shape

**Comment by iamemilio** [comment]
This is ready for review, and has been squashed down into more digestible commits.

File: parser/agent.go [hunk 1]
```diff
@@ -198,17 +228,23 @@ func NoticeError(manager *InstrumentationManager, stmt dst.Stmt, c *dstutil.Curs
...
-
+				comment.Warn(pkg, stmt, "Unchecked Error, please consult New Relic documentation on error capture")
 				manager.errorCache.Clear()
-				return true
-			} else {
-				manager.errorCache.Load(errExpr, nodeVal)
 			}
+
+			// Always load the error into the cache
+			fmt.Println("loaded into cache: ", errExpr)
```
Comment (mirackara) [inline parser/agent.go]: Leftover print statement

**Review by mirackara — COMMENTED** [review]
LGTM -- Just one leftover debug line

**Review by mirackara — APPROVED** [review]

**Pull request merged**